### DISCUSSION TITLE
After using setValue(), the updated field doesn't appear in submit data. What am I missing?

### DISCUSSION THREAD

**Discussion by domarp-j** [react-hook-form/react-hook-form#6913] [Q&A] [answered]
I'm building out a form where a user provides some information about themselves. There's a possibility that some of the user's information is in a Redux store, so I'm fetching that data and looking to initialize the form with this info. 

I'm currently doing the following: 

```js
const MyForm = () => {
  const {
    formState: { errors },
    handleSubmit,
    register,
    setValue,
    watch,
  } = useForm()

  // Get the user's information from the Redux store. 
  const user = useSelector((state) => state.user);

  // If user is defined, use setValue() to initialize some registered fields. 
  useEffect(() => {
    if (!user) return;

    setValue('email', user.email);
    setValue('firstName', user.firstName);
    setValue('lastName', user.lastName);
  }, [user]);

  // Submit handler
  const onSubmit = ({ email, firstName, lastName} ) => {
    JSON.stringify({
      email, 
      firstName, 
      lastName,
    }); 
  }

  // The rendered form (partial)
  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {...}
      <Input
        type="email"
        label="*Email Address"
        disabled={user.email}
        error={errorMessage({ field: 'email', errors })}
        {...register('email', { required })}
      />

      <Input
        label="*First Name"
        disabled={user.firstName}
        error={errorMessage({ field: 'firstName', errors })}
        {...register('firstName', { required })}
      />

      <Input
        label="*Last Name"
        disabled={user.lastName}
        error={errorMessage({ field: 'lastName', errors })}
        {...register('lastName', { required })}
      />
      {...}
    </>
  )
}
```

If `user` is undefined (not in Redux state), then a person can manually fill out their email, first name, and last name. On submit, I see an alert with their inputted information as expected. Everything's good so far. 

However, if `user` is defined (in the Redux store), then that `useEffect` call triggers three `setValue`s to initialize the value of email, firstName, and lastName. Additionally, the fields are disabled so that users can't manually change these values. The form renders properly with the initial values and disabled fields. However, **on submit, the email, firstName, and lastName arguments are always undefined.** I've had to fall back to the following to make this work: 

```js
const onSubmit = ({ email, firstName, lastName} ) => {
  JSON.stringify({
    // Email is defined if setValue() was never used.
    // If setValue() was used, email is undefined - I have to fall back to the user object's email. 
    email: email || user.email, 
    // Ditto for the remaining fields.
    firstName: firstName || user.firstName, 
    lastName: lastName || user.lastName,
  }); 
}
```

I'm sure I'm missing something here, but I'm not sure what. Any help would be appreciated!

EDIT: As @bluebill1049 states below, `disabled` is not the appropriate attribute here. `readOnly` is much more appropriate, and it passes along the field data to `onSubmit`.

**Comment by bluebill1049** [comment, marked as answer]
can you provide a codesandbox? also, why are you doing that in `useEfect` with the user as dep, it may fire multiple times?

**Reply by domarp-j** [reply]
Thanks for responding! 

I created a sandbox that's conceptually similar: https://codesandbox.io/s/gifted-ishizaka-vvny3?file=/src/App.js. The `user` is fetched asynchronously similar to my use case. However, that sandbox seems to be working as expected...

Could it be my custom `Input` component is the issue? While the CodeSandbox is working, when I replace the code above with `<input>`s rather than `Input` I'm still seeing undefined values for `email`, `firstName`, and `lastName` in `onSubmit`, so it seems unlikely.

The `Input` component definition is below for reference. 

```js
/**
 * **************************************************
 *
 * @module Input
 *
 * @description
 * Input is a standard text input that is fully
 * compatible with React Hook Form.
 *
 * **************************************************
 */

import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';

import styles from './Input.scss';

/**
 * ====================================
 *        COMPONENT
 * ====================================
 */

const Input = forwardRef(
  (
    {
      className,
      error,
      instructions,
      label,
      name,
      onChange,
      theme,
      type,
      ...rest
    },
    ref
  ) => {
    return (
      <div
        className={`${styles[theme]} ${
          error ? styles['with-errors'] : ''
        } ${className}`}
      >
        {label && <label htmlFor={name}>{label}</label>}

        {instructions && (
          <small className={styles.instructions}>{instructions}</small>
        )}

        <input
          id={name}
          name={name}
          onChange={onChange}
          ref={ref}
          type={type}
          {...rest}
        />

        {error && <div className={styles['error-message']}>{error}</div>}
      </div>
    );
  }
);

/**
 * ====================================
 *        METADATA
 * ====================================
 */

Input.displayName = 'Input';

Input.propTypes = {
  className: PropTypes.string,
  error: PropTypes.string,
  instructions: PropTypes.string,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  theme: PropTypes.oneOf(['default']),
  type: PropTypes.string,
};

Input.defaultProps = {
  className: '',
  theme: 'default',
  type: 'text',
};

/**
 * ====================================
 *        EXPORTS
 * ====================================
 */

export default Input;
```

**Reply by domarp-j** [reply]
After further testing, it looks like the `disabled` fields in the `Inputs` above are the culprit. When I remove the `disabled` prop from all `Inputs`, the `onSubmit` works perfectly and receives the `email`, `firstName`, and `lastName` values. 

Is there any way to still use `setValue()` on disabled fields?

**Reply by bluebill1049** [reply]
you want to use `readOnly` instead of `disabled`.

**Reply by domarp-j** [reply]
Thank you! I apologize for my confusion between `disabled` and `readOnly` here. Converting to `readOnly` worked like a charm.
